### ISSUE TITLE
Proliferation of forecasting base classes?

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/sktime#510]
@mloning, @big-o (creditef as contributor in the base classes):

I wanted to work on #464 and looked at the base classes in forecasting, amongst others.
I looked at the base classes and feel there is some unnecessary proliferation of classes going on.

Why: we seem to have two nested template base classes, `BaseForecaster` and `_SktimeForecaster`. 

The (undocumented) extension pattern seems unnecessarily complicated.
To give one example, implementation of prediction (can you confirm that this is the intended pattern, see new post in #464):

The `BaseForecaster` expects from extender to inherit and implement `predict`. 
However, if you implement a sktime native forecaster (which most extenders will be concerned with), you have to inherit from `_SktimeForecaster`, which inherits `BaseForecaster` and implements `predict`, but now you have to implement `_predict`.

I would strongly favour there being only one base class - with all private functions - and a clean docstring that states what to implement.
Only vendor adaptors and similar overrides should have nested template schemes.
Mixins (like for the forecasting horizon) are ok i.m.o.
That way, for example, one wouldn't have to track overrides through two parent classes...

What do you think, @mloning, @big-o? Next major version to refactor this?
Considering use of `abc` since it's getting a little complex?

**Comment by tasmn** [comment]
I think my contribution to those classes was when I tidied up the prediction intervals API, so I only have one very specific perspective on this. That said, I felt very much the same at the time I worked on it - I was unsure what the justification was for two separate classes (there was probably a good reason at the time, but it seems less clear now). It was difficult to figure out where new methods should live.

Something I liked in earlier versions of sktime that perhaps we could return to in this refactor is a simplification of the docstrings. At the moment, the documentation is spread across both private and public methods. Not only does this clutter up the codebase a little (and make it less readable), but it doesn't play as nicely with documentation tools like sphinx. I think it could be better to keep all of the meaningful user-focussed documentation in the public methods, and the private methods can contain very light documentation notes for developers (if absolutely necessary), or be undocumented (comments where they are called in the public methods can explain what they are being used for if needed). For example:

```python
def predict(self, fh, X=None, return_pred_int=False, alpha=DEFAULT_ALPHA):
    """Meaningful, generic, napoleon-formatted user documentation."""
    # Call class-specific implementation.
    return self._predict(fh=fh, X=X, return_pred_int=return_pred_int, alpha=alpha)

@abstractmethod
def _predict(self, fh, X=None, return_pred_int=False, alpha=DEFAULT_ALPHA):
    "A light, informal note *iff* it's useful for developers."
    raise NotImplementedError()
```

**Comment by big-o** [comment]
I think my contribution to those classes was when I tidied up the prediction intervals API, so I only have one very specific perspective on this. That said, I felt very much the same as you at the time I worked on it - I was unsure what the justification was for two separate classes (there was probably a good reason at the time, but it seems less clear now). It was difficult to figure out where new methods should live.

Something I liked in earlier versions of sktime that perhaps we could return to in this refactor is a simplification of the docstrings. At the moment, the documentation is spread across both private and public methods. Not only does this clutter up the codebase a little (and make it less readable), but it doesn't play as nicely with documentation tools like sphinx. I think it could be better to keep all of the meaningful user-focussed documentation in the public methods, and the private methods can contain very light documentation notes for developers (if absolutely necessary), or be undocumented (comments where they are called in the public methods can explain what they are being used for if needed). For example:

```python
def predict(self, fh, X=None, return_pred_int=False, alpha=DEFAULT_ALPHA):
    """Meaningful, generic, napoleon-formatted user documentation."""
    # Call class-specific implementation.
    return self._predict(fh=fh, X=X, return_pred_int=return_pred_int, alpha=alpha)

@abstractmethod
def _predict(self, fh, X=None, return_pred_int=False, alpha=DEFAULT_ALPHA):
    "A light, informal note *iff* it's useful for developers."
    raise NotImplementedError()
```

**Comment by fkiraly** [comment]
>I think it could be better to keep all of the meaningful user-focussed documentation in the public methods, and the private methods can contain very light documentation notes for developers (if absolutely necessary), or be undocumented (comments where they are called in the public methods can explain what they are being used for if needed).

Hm, I don't like that - I think even abstracts and privates should be well-documented in-place, i.e., directly in the code.
More docstrings do not necessarily mean more readability, but in this case (and more often than not) I think it does - as the base classes would be a common place for advanced users to look for an explanation of how sktime works internally.

**Comment by mloning** [comment]
Hi @fkiraly, yes I agree, the forecasting API needs to be simplified, this is still mostly my first attempt. 

I'm not sure what the best solution is in terms of base classes. Many implementations at the moment are in fact interfaces to other packages. I agree that the use of `_predict` saves us from some code duplication but makes it hard to understand and extend.

In hindsight, I'd also remove the `update_predict` functionality from the estimator class and implement in a separate stand-alone function. It adds a lot of complications inside the class for handling state changes.

**Comment by fkiraly** [comment]
High-level, I see two major options - both resulting in a merge of the two base class templates:
* make it more lean and adopt the pattern where `fit`, `predict`, etc are overridden in concrete implementation descendants
* make it nested and have `fit` and `_fit` and `predict` and `_predict`, where utility functionality and other plumbing sits in the public methods, while extenders consistently implement the private ones

An "outsider" option but perhaps more elegant is to do the first, and mimick behaviour of the second via decorators on the methods.

**Comment by fkiraly** [comment]
>In hindsight, I'd also remove the update_predict functionality from the estimator class and implement in a separate stand-alone function. It adds a lot of complications inside the class for handling state changes.

Regarding `update_predict`: I'd advocate to leave it in, because it's a very common deployment mode that users might expect and actually use (I suspect without concrete evidence but from my view of the field - opinions anyone?).
We just need to very carefully think about the state model here.

**Comment by mloning** [comment]
In deployment, you'd use `update` and `predict` separately or what we call `update_predict_single` - of course they need to stag. 

But `update_predict` does temporal CV which creates lots of complications - I don't see any reasons why that should live in the class but open to other thoughts?

**Comment by mloning** [comment]
From #941 on the refactoring of `_set_fh`:

I've thought about this a bit more and I prefer the option where we pass the name of the calling method to _set_fh, e.g. _set_fh(fh, method="fit"). For me, the main reasons are:

* it keeps the functionality for checking/setting the fh contained to that function and does not assume state changes elsewhere in the code, like at the beginning of fit,
* it keeps it closer to the implementation of other estimator types where we do not re-set the state at the beginning of fit,
* it avoids the redundancy of having to set _is_fitted=False in the constructor and at the beginning of fit.

**Comment by fkiraly** [comment]
solved now via #912
